
<style>
.jp-Notebook .jp-MarkdownCell,
.text_cell_render {
    font-size: 14px;
    line-height: 1.38;
}
.jp-CodeCell .cm-editor,
.jp-CodeCell pre,
.code_cell .CodeMirror {
    font-size: 13px !important;
    line-height: 1.30 !important;
}
</style>

# ECON 422: Macroeconomics & Machine Learning
## Hyperparameters — Coding Practice

For the loss function

$$
L_a(\theta)
=
\sum_{i=1}^{n}
\left|y_i-\theta\right|^a,
$$

- when $a=1$, the fitted parameter is the sample **median**;
- when $a=2$, the fitted parameter is the sample **mean**.

We use two-fold cross-validation to compare the two hyperparameter values.


### Setup

In [ ]:
import numpy as np

### Practice 1 — Prepare the data and folds

Type the gray code in the blank cell below.

<pre style="color:#8a8a8a;background:#f7f7f7;border:1px solid #e4e4e4;border-radius:5px;padding:9px 11px;margin:6px 0 8px 0;font-family:Menlo, Monaco, Consolas, 'Courier New', monospace;font-size:12.5px;line-height:1.32;white-space:pre;overflow-x:auto;">y = np.array([1, 2, 10, 1, 2, 3], dtype=float)

folds = [
    np.array([0, 1, 2]),
    np.array([3, 4, 5])
]</pre>

In [ ]:
# Type Practice 1 here.
y = np.array([1, 2, 10, 1, 2, 3], dtype=float)

folds = [
    np.array([0, 1, 2]),
    np.array([3, 4, 5])
]

### Practice 2 — Cross-validation for $a=1$

For $a=1$, estimate $\theta$ by the **median** in each estimation fold.

<pre style="color:#8a8a8a;background:#f7f7f7;border:1px solid #e4e4e4;border-radius:5px;padding:9px 11px;margin:6px 0 8px 0;font-family:Menlo, Monaco, Consolas, 'Courier New', monospace;font-size:12.5px;line-height:1.32;white-space:pre;overflow-x:auto;">losses_a1 = []

for val_idx in folds:
    train_idx = np.setdiff1d(np.arange(len(y)), val_idx)
    y_train, y_val = y[train_idx], y[val_idx]

    theta = np.median(y_train)
    losses_a1.append(np.mean((y_val - theta)**2))

cv_a1 = np.mean(losses_a1)</pre>

In [3]:
# Type Practice 2 here.
losses_a1 = []

for val_idx in folds:
    train_idx = np.setdiff1d(np.arange(len(y)), val_idx)
    y_train, y_val = y[train_idx], y[val_idx]
    
    theta = np.median(y_train)
    losses_a1.append(np.mean((y_val - theta) ** 2))

cv_a1 = np.mean(losses_a1)

### Practice 3 — Cross-validation for $a=2$

For $a=2$, estimate $\theta$ by the **mean** in each estimation fold.

<pre style="color:#8a8a8a;background:#f7f7f7;border:1px solid #e4e4e4;border-radius:5px;padding:9px 11px;margin:6px 0 8px 0;font-family:Menlo, Monaco, Consolas, 'Courier New', monospace;font-size:12.5px;line-height:1.32;white-space:pre;overflow-x:auto;">losses_a2 = []

for val_idx in folds:
    train_idx = np.setdiff1d(np.arange(len(y)), val_idx)
    y_train, y_val = y[train_idx], y[val_idx]

    theta = np.mean(y_train)
    losses_a2.append(np.mean((y_val - theta)**2))

cv_a2 = np.mean(losses_a2)</pre>

In [6]:
# Type Practice 3 here.
losses_a2 = []


for val_idx in folds:
    train_idx = np.setdiff1d(np.arange(len(y)), val_idx)
    y_train, y_val = y[train_idx], y[val_idx]
    
    theta = np.mean(y_train)
    losses_a2.append(np.mean((y_val - theta) ** 2))

cv_a2 = np.mean(losses_a2)

### Practice 4 — Select the hyperparameter

Compare

$$
CV(1)
\quad\text{and}\quad
CV(2),
$$

and choose the value of $a$ with the smaller average validation loss.

<pre style="color:#8a8a8a;background:#f7f7f7;border:1px solid #e4e4e4;border-radius:5px;padding:9px 11px;margin:6px 0 8px 0;font-family:Menlo, Monaco, Consolas, 'Courier New', monospace;font-size:12.5px;line-height:1.32;white-space:pre;overflow-x:auto;">cv = {1: cv_a1, 2: cv_a2}
best_a = min(cv, key=cv.get)

print(&quot;CV:&quot;, cv)
print(&quot;Selected a:&quot;, best_a)</pre>

In [8]:
# Type Practice 4 here.

cv = {1: cv_a1, 2: cv_a2}
best_a = min(cv, key=cv.get)

print("CV:", cv)
print("Selected a:", best_a)

CV: {1: np.float64(11.166666666666668), 2: np.float64(13.88888888888889)}
Selected a: 1


### Practice 5 — Refit using the full sample

After selecting $\widehat{a}$, estimate the final parameter using all observations:

$$
\widehat{\theta}
=
\arg\min_{\theta}
\sum_{i=1}^{n}
\left|y_i-\theta\right|^{\widehat{a}}.
$$

<pre style="color:#8a8a8a;background:#f7f7f7;border:1px solid #e4e4e4;border-radius:5px;padding:9px 11px;margin:6px 0 8px 0;font-family:Menlo, Monaco, Consolas, 'Courier New', monospace;font-size:12.5px;line-height:1.32;white-space:pre;overflow-x:auto;">theta_final = (
    np.median(y) if best_a == 1
    else np.mean(y)
)

print(&quot;Final theta:&quot;, theta_final)</pre>

In [10]:
# Type Practice 5 here.
theta_final = (
    np.median(y) if best_a == 1
    else np.mean(y)
)

print("Final theta:", theta_final)

Final theta: 2.0



### Result

The cross-validation scores are approximately

$$
CV(1) \approx 11.17,
\qquad
CV(2) \approx 13.89.
$$

Therefore,

$$
\widehat a = 1,
\qquad
\widehat\theta = 2.
$$



### Exercise

Copy and paste data-generating-process codes from '(Sep2) Exercise' on Canvas.

Use

$$
n=40
$$

and random seed `7`. Randomly divide the sample into two folds. Then use two-fold cross-validation with mean squared validation loss to choose between

$$
a\in\{1,2\}.
$$

Finally, refit $\theta$ on the full sample using the selected value of $a$.


The four blanks correspond to:

1. the estimator when $a=1$;
2. the estimator when $a=2$;
3. the validation mean squared error; and
4. selecting the value of $a$ with the smaller cross-validation loss.

In [11]:
# Copy and paste dgp codes from '(Sep2) Exercise' on Canvas.

rng = np.random.default_rng(7)
n = 40

# Generate the DGP
eps = rng.normal(0, 1, n)
D = rng.binomial(1, 0.10, n)
y_ex = 2 + eps + 8 * D

print("Number of contaminated observations:", D.sum())

# Create two random folds
idx = rng.permutation(n)
folds_ex = [
    idx[:n // 2],
    idx[n // 2:]
]

# Cross-validation
cv_ex = {}

for a in [1, 2]:
    losses = []

    for val_idx in folds_ex:
        train_idx = np.setdiff1d(np.arange(n), val_idx)

        y_train = y_ex[train_idx]
        y_val = y_ex[val_idx]

        theta = np.median(y_train) if a == 1 else np.mean(y_train)
        losses.append(np.mean((y_val - theta) ** 2))

    cv_ex[a] = np.mean(losses)

best_a_ex = min(cv_ex, key=cv_ex.get)

# Refit on the full sample
theta_final_ex = (
    np.median(y_ex) if best_a_ex == 1
    else np.mean(y_ex)
)

print("CV:", cv_ex)
print("Selected a:", best_a_ex)
print("Final theta:", theta_final_ex)


Number of contaminated observations: 4
CV: {1: np.float64(6.908950711532283), 2: np.float64(7.2481645478590355)}
Selected a: 1
Final theta: 1.850683552892943
